### PR TITLE
Adddgetri

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
This needs https://github.com/gonum/lapack/pull/52 .

The cgo lapack implementation is wrong. The function documents that it returns the matrix inverse, but the cgo implementation returns a permutation of the inverse.

File: native/dgetri.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+// Dgetri computes the inverse of the matrix A using the LU factorization computed
+// by Dgetrf. On entry, a contains the PLU decomposition of A as computed by
+// Dgetrf and on exit contains the reciprocal of the original matrix.
+//
+// Work is temporary storage, and lwork specifies the usable memory length.
+// At minimum, lwork >= n and this function will panic otherwise.
+// Dgetri is a blocked inversion, but the block size is limited
+// by the temporary space available. If lwork == -1, instead of performing Dgetri,
+// the optimal work length will be stored into work[0].
+func (impl Implementation) Dgetri(n int, a []float64, lda int, ipiv []int, work []float64, lwork int) bool {
```
Comment (kortschak) [inline native/dgetri.go]: The return is not described.

File: native/dgetri.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	nbmin := 2
+	ldwork := nb
+	if nb > 1 && nb < n {
+		iws := max(ldwork*n, 1)
+		if lwork < iws {
+			nb = lwork / ldwork
+			nbmin = max(2, impl.Ilaenv(2, "DGETRI", " ", n, -1, -1, -1))
+		}
+	}
+	ldwork = nb
```
Comment (kortschak) [inline native/dgetri.go]: Why?

Comment (btracey) [inline native/dgetri.go]: I can leave a note. It's a column major vs. row major thing. Work is effectively an n x nb matrix, so ldwork is nb and not n. Either way this particular line should be deleted.

File: native/dgetri.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	bi := blas64.Implementation()
+	// TODO(btracey): Replace this with a more row-major oriented algorithm.
+	if nb < nbmin || nb >= n {
+		// Unblocked code.
+		for j := n - 1; j >= 0; j-- {
+			for i := j + 1; i < n; i++ {
+				work[i*ldwork] = a[i*lda+j]
```
Comment (kortschak) [inline native/dgetri.go]: Why stride this by ldwork?

Comment (btracey) [inline native/dgetri.go]: I haven't transposed work. Before, the indexing is effectively work(i,1), and I have transformed that indexing into row major. this particular loop is easy to re-orient, but it's more complicated below in the Dgemv and Dgemm calls.

File: native/dgetri.go (hunk 1 above)
Comment (btracey) [inline native/dgetri.go]: Done.

File: native/dgetri.go (hunk 3 above)
Comment (kortschak) [inline native/dgetri.go]: Yes, sorry, that's what I mean.

File: cgo/lapack.go [hunk 4]
```diff
@@ -395,6 +395,38 @@ func (impl Implementation) Dgetrf(m, n int, a []float64, lda int, ipiv []int) (o
...
+// by Dgetrf. On entry, a contains the PLU decomposition of A as computed by
+// Dgetrf and on exit contains the reciprocal of the original matrix.
+//
+// Dtrti returns whether the matrix a is singular or whether it's not singular.
+// If the matrix is singular the inversion is not performed.
+//
+// The C interface does not support providing temporary storage. To provide compatibility
+// with native, lwork == -1 will not run Dgetri but will instead write the minimum
+// work necessary to work[0]. If len(work) < lwork, Dgetri will panic.
+func (impl Implementation) Dgetri(n int, a []float64, lda int, ipiv []int, work []float64, lwork int) bool {
```
Comment (kortschak) [inline cgo/lapack.go]: Make this (ok bool) to match the native method.

Comment (btracey) [inline cgo/lapack.go]: Done.

**Comment by btracey** [comment]
Note that the cgo error is now logged (with comment).

File: native/dgetri.go (hunk 1 above)
Comment (kortschak) [inline native/dgetri.go]: "ok" here too.

**Comment by kortschak** [comment]
Here we go again:

```
$ go test github.com/gonum/lapack/cgo -cpu 2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2 -run Dgetri
fatal error: unexpected signal during runtime execution
[signal 0xb code=0x1 addr=0x0 pc=0x7f66fb5a3e2c]

runtime stack:
runtime.throw(0x66f060, 0x2a)
	/home/daniel/Development/go/src/runtime/panic.go:527 +0x90
runtime.sigpanic()
	/home/daniel/Development/go/src/runtime/sigpanic_unix.go:12 +0x5a

goroutine 5 [syscall, locked to thread]:
runtime.cgocall(0x4280c0, 0xc820043b18, 0xc800000000)
	/home/daniel/Development/go/src/runtime/cgocall.go:120 +0x11b fp=0xc820043ad0 sp=0xc820043aa0
github.com/gonum/lapack/cgo/clapack._Cfunc_LAPACKE_dgetri(0x500000065, 0xc820067040, 0x5, 0xc820010420, 0xc800000000)
	github.com/gonum/lapack/cgo/clapack/_obj/_cgo_gotypes.go:6049 +0x39 fp=0xc820043b18 sp=0xc820043ad0
github.com/gonum/lapack/cgo/clapack.Dgetri(0x5, 0xc820067040, 0x19, 0x19, 0x5, 0xc820010420, 0x5, 0x5, 0x5)
	/home/daniel/Development/src/github.com/gonum/lapack/cgo/clapack/clapack.go:1192 +0x59 fp=0xc820043b48 sp=0xc820043b18
github.com/gonum/lapack/cgo.Implementation.Dgetri(0x5, 0xc820067040, 0x19, 0x19, 0x5, 0xc82000a540, 0x5, 0x5, 0xc82000a570, 0x5, ...)
	/home/daniel/Development/src/github.com/gonum/lapack/cgo/lapack.go:427 +0x27f fp=0xc820043bc0 sp=0xc820043b48
github.com/gonum/lapack/cgo.(*Implementation).Dgetri(0x966150, 0x5, 0xc820067040, 0x19, 0x19, 0x5, 0xc82000a540, 0x5, 0x5, 0xc82000a570, ...)
	<autogenerated>:13 +0x12a fp=0xc820043c30 sp=0xc820043bc0
github.com/gonum/lapack/testlapack.DgetriTest(0xc8200a2000, 0x7f66f2a6f798, 0x966150)
	/home/daniel/Development/src/github.com/gonum/lapack/testlapack/dgetri.go:57 +0x620 fp=0xc820043f28 sp=0xc820043c30
github.com/gonum/lapack/cgo.TestDgetri(0xc8200a2000)
	/home/daniel/Development/src/github.com/gonum/lapack/cgo/lapack_test.go:61 +0x76 fp=0xc820043f68 sp=0xc820043f28
testing.tRunner(0xc8200a2000, 0x930ea8)
	/home/daniel/Development/go/src/testing/testing.go:456 +0x98 fp=0xc820043fa0 sp=0xc820043f68
runtime.goexit()
	/home/daniel/Development/go/src/runtime/asm_amd64.s:1696 +0x1 fp=0xc820043fa8 sp=0xc820043fa0
created by testing.RunTests
	/home/daniel/Development/go/src/testing/testing.go:561 +0x86d

goroutine 1 [chan receive]:
testing.RunTests(0x686be8, 0x930da0, 0x11, 0x11, 0x611c01)
	/home/daniel/Development/go/src/testing/testing.go:562 +0x8ad
testing.(*M).Run(0xc820053ef8, 0x43a9e8)
	/home/daniel/Development/go/src/testing/testing.go:494 +0x70
main.main()
	github.com/gonum/lapack/cgo/_test/_testmain.go:86 +0x116

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/home/daniel/Development/go/src/runtime/asm_amd64.s:1696 +0x1
FAIL	github.com/gonum/lapack/cgo	0.008s
```

This is abolished by commenting out the call the LAPACK_dgetri, so the prep is incorrect for that routine. I think we need Indiana Jones for this.

File: cgo/lapack.go [hunk 5]
```diff
@@ -395,6 +395,38 @@ func (impl Implementation) Dgetrf(m, n int, a []float64, lda int, ipiv []int) (o
...
+	}
+	if lwork < n {
+		panic(badWork)
+	}
+	if len(work) < lwork {
+		panic(badWork)
+	}
+	ipiv32 := make([]int32, len(ipiv))
+	for i, v := range ipiv {
+		ipiv32[i] = int32(v)
```
Comment (kortschak) [inline cgo/lapack.go]: ipiv32[i] = int32(v) + 1

**Comment by btracey** [comment]
Ah shoot.

File: native/dgetri.go (hunk 1 above)
Comment (btracey) [inline native/dgetri.go]: Done.

**Comment by kortschak** [comment]
LGTM after minor change. Any idea why travis fails this? The usual?

**Pull request merged**